File: julien-lectures/CODE/Program_6_6.m
function [t,X,Y,Z] = Program_6_4(beta,gamma,mu,epsilon,delta,X0,Y0,N0,MaxTime)
%
% 
% Program_6_4( beta, gamma, mu, epsilon, delta, X0, Y0, N0, MaxTime)
%      This is the MATLAB version of program 6.6 from page 210 of 
% "Modeling Infectious Disease in humans and animals" 
% by Keeling & Rohani.
% 
% It is the SIR model (including births and deaths) with full 
% (event-driven) demographic stochasticity. Two forms of imports (governed
% by the parameters epsilon and delta) are implemented.
%
% This is a more complex stochastic model as 8 events are
% possible: infection, recovery, birth, death of susceptible, death of
% infected, death of recovered and two forms of imports
%
% Note: by default we are using a very small population size to highlight
% the stochasticity.

% Sets up default parameters if necessary.
if nargin == 0
   beta=1;
   gamma=1/10.0;
   mu=5e-4;
   N0=5000;
   delta=0.01;
   epsilon=delta*10/N0;
   
   
   Y0=ceil(mu*N0/gamma);
   X0=floor(gamma*N0/beta);
   
   MaxTime=10*365;
end

% Checks all the parameters are valid
if X0<=0 
    error('Initial number of susceptibles (%g) is less than or equal to zero',X0);
end

if Y0<=0 
    error('Initial number of infecteds (%g) is less than or equal to zero',Y0);
end

if N0<=0 
    error('Initial population size (%g) is less than or equal to zero',N0);
end

if beta<=0 
    error('Transmission rate beta0 (%g) is less than or equal to zero',beta0);
end

if gamma<=0 
    error('Recovery rate gamma (%g) is less than or equal to zero',gamma);
end

if mu<=0 
    error('Brith and Death rate mu (%g) is less than or equal to zero',mu);
end

if epsilon<0 
    error('Import rate epsilon (%g) is less than zero',epsilon);
end

if delta<0 
    error('Import rate delta (%g) is less than zero',delta);
end
  
if MaxTime<=0 
    error('Maximum run time (%g) is less than or equal to zero',MaxTime);
end

if X0+Y0>N0
    warning('Initial level of suceptibles + infecteds (%g + %g = %g) is greater than population size (%g)',X0,Y0,X0+Y0,N0);
end


Z0=N0-X0-Y0;

% The main iteration 

[t, epsilonT, deltaT, pop]=Stoch_Iteration([0 MaxTime],[X0 Y0 Z0],[beta gamma mu epsilon delta N0]);

T=t/365; X=pop(:,1); Y=pop(:,2); Z=pop(:,3);

% plots the graphs with scaled colours
subplot(2,1,1)
h=plot(T,X,'-g');
set(gca,'XLim',[0 max(T)]);
xlabel 'Time (years)';
ylabel 'Susceptible'

subplot(2,1,2)
h=plot(T,Y,'-r');
Q=get(gca,'YLim');
hold on
m=find(epsilonT>0);
h1=plot([-1000 T(m)],0,'ok'); set(h1,'MarkerFaceColor','b','MarkerSize',4);
m=find(deltaT>0);
h2=plot([-1000 T(m)],Q(2),'ok'); set(h2,'MarkerFaceColor','r','MarkerSize',4);
hold off
axis([0 max(T) 0 Q(2)*1.05]);
legend([h1(1) h2(1)],'Epsilon Imports','Delta Imports');
xlabel 'Time (years)';
ylabel 'Infectious'





% Do the iterations using the full event driven stochastic methodology
% This is a relatively general version of the event-driven methodology know
% as Gillespie's Direct Algorithm
function [T,ET,DT,P]=Stoch_Iteration(Time,Initial,Parameters)

X=Initial(1); Y=Initial(2); Z=Initial(3);
T=0; ET=0; DT=0; P(1,:)=[X Y Z];
old=[X Y Z];

loop=1;
while (T(loop)<Time(2))  
    [step,et,dt,new]=Iterate(old,Parameters);
    loop=loop+1;
    T(loop)=T(loop-1)+step;
    P(loop,:)=old;
    loop=loop+1;
    T(loop)=T(loop-1);
    P(loop,:)=new; old=new;
    ET(loop)=et; DT(loop)=dt;
    
    if loop>=length(T)
        T(loop*2)=0;
        P(loop*2,:)=0;
        ET(loop*2)=0;
        DT(loop*2)=0;
    end
end

T=T(1:loop); P=P(1:loop,:); ET=ET(1:loop); DT=DT(1:loop);


% Do the actual iteration step
function [step, et, dt, new_value]=Iterate(old, Parameters);

beta=Parameters(1); gamma=Parameters(2); mu=Parameters(3); 
epsilon=Parameters(4); delta=Parameters(5);
N=Parameters(6);
X=old(1); Y=old(2); Z=old(3);

Rate(1) = beta*X*Y/N;  Change(1,:)=[-1 +1 0];
Rate(2) = gamma*Y;  Change(2,:)=[0 -1 +1];
Rate(3) = mu*N;  Change(3,:)=[+1 0 0];
Rate(4) = mu*X;  Change(4,:)=[-1 0 0];
Rate(5) = mu*Y;  Change(5,:)=[0 -1 0];
Rate(6) = mu*Z;  Change(6,:)=[0 0 -1];
Rate(7) = epsilon*X;  Change(7,:)=[-1 +1 0];
Rate(8) = delta;  Change(8,:)=[0 +1 0];

R1=rand(1,1);
R2=rand(1,1);

step = -log(R2)/(sum(Rate));

% find which event to do
m=min(find(cumsum(Rate)>=R1*sum(Rate)));

new_value=old+Change(m,:);

dt=0; et=0;

if m==7
    et=1;
end
if m==8
    dt=1;
end

File: julien-lectures/CODE/Program_6_6.py
#!/usr/bin/env python

####################################################################
###    This is the PYTHON version of program 6.4 from page 203 of  #
### "Modeling Infectious Disease in humans and animals"            #
### by Keeling & Rohani.										   #
###																   #
### It is the SIR model (including births and deaths) with full    #
### (event-driven) demographic stochasticity. Two forms of imports #
### (governed by the parameters epsilon and delta) are implemented.#
###																   #
### This is a more complex stochastic model as 8 events are		   #
### possible: infection, recovery, birth, death of susceptible,    #
### death of infected, death of recoveredand two forms of imports  #
###																   #
### Note: by default we are using a very small population size 	   #
### to highlight the stochasticity.				   				   #
####################################################################

###################################
### Written by Ilias Soumpasis    #
### ilias.soumpasis@ucd.ie (work) #
### ilias.soumpasis@gmail.com	  #
###################################

import numpy as np
import pylab as pl
from matplotlib.font_manager import FontProperties as mfp

beta=1.0
gamma=1/10.0
mu=5e-4
N0=5000.0
### You may want to try with popylation size of 50 (small) to see the events
### In this case uncomment the next line
#N0=50.0
delta=0.01;
epsilon=delta*10/N0;
ND=MaxTime=10*365.0
Y0=pl.ceil(mu*N0/gamma)
X0=pl.floor(gamma*N0/beta)
Z0=N0-X0-Y0

INPUT = np.array((X0,Y0,Z0, 0.0, 0.0))

timestep=0.0

def stoc_eqs(INP,ts): 
	V = INP
	Rate=np.zeros((8))
	Change=np.zeros((8,3))
	N=pl.sum(V[range(3)])
	Rate[0] = beta*V[0]*V[1]/N;  Change[0,:]=([-1, +1, 0]);
	Rate[1] = gamma*V[1];  Change[1,:]=([0, -1, +1]);
	Rate[2] = mu*N;  Change[2,:]=([+1, 0, 0]);
	Rate[3] = mu*V[0];  Change[3,:]=([-1, 0, 0]);
	Rate[4] = mu*V[1];  Change[4,:]=([0, -1, 0]);
	Rate[5] = mu*V[2];  Change[5,:]=([0, 0, -1]);
	Rate[6] = epsilon*V[0];  Change[6,:]=[-1, +1, 0];
	Rate[7] = delta;  Change[7,:]=[0, +1, 0];
	R1=pl.rand();
	R2=pl.rand();
	ts = -np.log(R2)/(np.sum(Rate));
	m=min(pl.find(pl.cumsum(Rate)>=R1*pl.sum(Rate)));
	V[range(3)]=V[range(3)]+Change[m,:]
	V[3]=0; V[4]=0.;
	if m == 6: V[3] = 1.
	if m == 7: V[4] = 1.
	return [V,ts]

def Stoch_Iteration(INPUT):
	lop=0
	ts=0
	T=[0]
	S=[0]
	I=[0]
	R=[0]
	et=[]
	dt=[]
	while T[lop] < ND:
		lop=lop+1
		T.append(T[lop-1]+ts)
		S.append(INPUT[0])
		I.append(INPUT[1])
		R.append(INPUT[2])
		et.append(INPUT[3])
		dt.append(INPUT[4])
		[res,ts] = stoc_eqs(INPUT,ts)
		lop=lop+1
		T.append(T[lop-1])
		S.append(INPUT[0])
		I.append(INPUT[1])
		R.append(INPUT[2])
		et.append(INPUT[3])
		dt.append(INPUT[4])
		if lop>len(T):
			T[lop*2]=0
			S[lop*2]=0
			I[lop*2]=0
			R[lop*2]=0
			et[lop*2]=0
			dt[lop*2]=0
	return [T,S,I,R, et, dt]

[T,S,I,R, et, dt]=Stoch_Iteration(INPUT)

tT=np.array(T)[1:,]/365.
tS=np.array(S)[1:,]
tI=np.array(I)[1:,]
tR=np.array(R)[1:,]
indpl=pl.find(tT<10)

### plotting
pl.subplot(211)
h=pl.plot(tT[pl.find(tT<10)], tS[pl.find(tT<10)], 'g')
#pl.xlabel ('Time (years)')
pl.ylabel ('Susceptible')
pl.subplot(212)
pl.plot(tT[pl.find(tT<10)], tI[pl.find(tT<10)], 'r')
ax=pl.gca()
me=pl.find(np.array(dt)>0)
md=pl.find(np.array(dt)>0)
pl.plot(tT[me],np.zeros((len(me)))+ax.get_ylim()[0]+5, 'bo', linewidth=4, label='Epsilon imports')
pl.plot(tT[me],np.zeros((len(me)))+ax.get_ylim()[1]-5, 'ro', linewidth=4, label='Delta imports')
pl.legend(loc=7,prop=mfp(size='smaller'))
pl.ylabel ('Infectious')
pl.show()
